pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/PendingAckHandleImpl.java

```diff
@@ -937,19 +937,24 @@ public TransactionPendingAckStats getStats(boolean lowWaterMarks) {
         return transactionPendingAckStats;
     }
 
-    public synchronized void completeHandleFuture() {
-        if (!this.pendingAckHandleCompletableFuture.isDone()) {
-            this.pendingAckHandleCompletableFuture.complete(PendingAckHandleImpl.this);
-        }
-        if (recoverTime.getRecoverStartTime() != 0L) {
-            recoverTime.setRecoverEndTime(System.currentTimeMillis());
+    public void completeHandleFuture() {
+        synchronized (this.pendingAckHandleCompletableFuture) {
```

Comment:
Why are you using synchronized on this variable ?
This smells a little bit (synchronize over a CompletableFuture)
And also that reference escapes this object because there is a getter, so I don't think we are going to properly deal with concurrency here

Reply:
the getter escapes on purpose but I agree we don't need the synchronized at all here

Reply:
do we need this at all (including `if (!this.pendingAckHandleCompletableFuture.isDone())` check)?
CompleteableFuture can be completed only once (API contract + implementation confirms this), if needed one can use return value of complete() to check if it succeeded

Reply:
updated, thanks